Custom loss function

peterewills opened the pull request:
Added support for custom loss functions in `TFFMClassifier`. Implemented cross-entropy in the `utils` module. Also added a `class_weight` parameter, which automatically uses weighted cross-entropy. One can use `class_weight = "balanced"` to use the heuristic `pos_weight = n_negative / n_positive`, where `n_positive` is the number of positive samples in the training labels.

Note that custom loss functions are disallowed for `TFFMRegressor` at the moment. Adding them would be trivial, but I don't know of a use case where you'd want something other than MSE, so I left it alone.

I will add a demo of this functionality to the example notebook sometime soon.

peterewills committed:
implemented weighted cross-entropy

peterewills committed:
updated regressor behavior

geffy commented on tffm/core.py:
```diff
@@ -91,22 +97,32 @@ class TFFMCore():
-                optimizer=tf.train.AdamOptimizer(learning_rate=0.01), reg=0, init_std=0.01, 
-                use_diag=False, reweight_reg=False, seed=None):
+                optimizer=tf.train.AdamOptimizer(learning_rate=0.01), reg=0,
+                init_std=0.01, class_weight=None, use_diag=False, reweight_reg=False,
```
IMHO, class_weight should be parameter of `TFFMClassifier.fit` method, not constructor of `Base`. (and can be changed during training without creating new class instance)

peterewills replied:
Agreed.

EDIT: As I think about this, I'm not so sure. Can you explain your reasoning here? For most algorithms in Scikit-Learn the loss function is a constructor of the class. I was mimicking that behavior, but if you have good reasons to go against it, I'm open to it.

geffy commented on tffm/core.py:
```diff
@@ -91,22 +97,32 @@ class TFFMCore():
         self.n_features = None
         self.graph = None
+        # loss is weighted cross entropy, if class_weight is given
+        if class_weight is not None:
```
It's a bad idea to have several almost equal loss functions. It's better to use single loss with a signature like (outputs, y, persample_weights) and calculate persample_weights in fit method.
This will require some additional code (like new placeholder for weights), but this approach allows to implement class weights and sample weights and doesn't require several loss_functions

peterewills replied:
Yeah it's true, I was just doing it the lazy way. I'll re-work this so that there's one `loss_xentropy` with a keyword arg of `pos_weight=1`, which just uses `tf.nn.weighted_cross_entropy_with_logits`. Then I'll use lambda functions, i.e.

`self.loss_function = lambda outputs, y: loss_xentropy(outputs, y, pos_weight = self.class_weights)`

Does that seem reasonable?

geffy commented on tffm/models.py:
```diff
@@ -25,13 +25,14 @@ class TFFMClassifier(TFFMBaseModel):
     def preprocess_target(self, y_):
         # suppose input {0, 1}, but internally will use {-1, 1} labels instead
-        assert(set(y_) == set([0, 1]))
+        if not (set(y_) == set([0, 1])):
```
Is there a reason to do check this way? 
I'm fully OK with error message, it's great. But why not to put it in `assert` instead of separate `if` and `raise`?

peterewills replied:
I find `AssertionError` to be ambiguous, I prefer to raise an error that explains the problem. That's a style issue though, if you prefer assertions that's OK too.

geffy commented on tffm/models.py:
```diff
@@ -81,12 +82,16 @@ class TFFMRegressor(TFFMBaseModel):
 
     def __init__(self, **init_params):
-        assert 'loss_function' not in init_params
+        # overwrite any custom loss function
```
Logic with overwriting is not transparent and we should avoid it, I believe.

peterewills replied:
I suppose the better solution would be to raise an error if `loss_function` or `class_weight` was in `init_params`. Similar to the last one, we can stick to assertions if you prefer, but I find them ambiguous.

geffy commented:
Hi @peterewills ! Thanks for contributing!
I've made several comments about implementation. Feel free to discuss if you are not agree with them. It will be great if you fix them. If not -- I believe I can do it by myself at the end of the week.

peterewills commented:
Made some comments. Let me know what you think, and I'll work on fixing these things end of this week or early next week.

peterewills committed:
renamed to pos_weight, removed balanced option

peterewills reviewed (commented):
I renamed class_weight to pos_weight, since only a single weight is provided. The `pos_weight="balanced"` option has been removed.

peterewills commented:
I was running into problems getting the "balanced" weight to be calculated over the whole training set rather than the mini-batch, so I just removed that option, forcing the user to set the weight directly. There is now only one additional loss function, `utils.loss_xentropy` which can be weighted with a keyword argument `pos_weight`. Moved the assignment of the loss function to the initializer for `TFFMClassifier`, rather than `TFFMCore`. Also, an error is raised if the user tries to provide a custom loss function to `TFFMRegressor`, making it more explicit that only MSE error can be used.

If you'd still like to have `pos_weight` as an parameter of `classifier.fit()`, then I'd be fine with it, but I think I'll leave it up to you.

What has me avoiding adding a `balanced` option is the complication of evaluating the positive weight. Would we have to put a placeholder in the graph, then assign it a value when `fit()` is called?

geffy committed:
improve sample_weight support

geffy commented:
I re-write some parts and add 'balanced' option. Can you, please, check it? Does it solve your original problem with sample weights?

peterewills commented:
It seems like you've written this to be very general, so that arbitrary weights can be provided for each sample. This is more general than I've ever needed to use, but I suppose it's fine. I don't like the disallowing of arbitrary loss functions for the `TFFMClassifier` class; I thought that was the whole point. Also, I think we should allow the user to provide the positive weight at the time of classifier construction if desired, a la scikit-learn. I'll implement these small changes when I get a chance.

Also not sure why `loss_xentropy` got removed.

peterewills commented on tffm/models.py:
```diff
@@ -25,14 +18,39 @@ class TFFMClassifier(TFFMBaseModel):
 
     def __init__(self, **init_params):
-        assert 'loss_function' not in init_params
+        assert 'loss_function' not in init_params, "Parameter 'loss_function' is unexpected for TFFMClassifier"
```
I thought the whole point was to allow for arbitrary custom loss functions for a `TFFMClassifier`. Why are we prohibiting this?

peterewills commented on tffm/utils.py:
```diff
@@ -207,4 +206,5 @@ def loss_logistic(outputs, y):
 
 def loss_mse(outputs, y):
     return tf.pow(y -  tf.transpose(outputs), 2, name='mse_loss')
+    
```
What happened to cross-entropy loss?

geffy commented:
Cross-entropy loss for binary classification is exactly a logloss. Since TFFM supports only binary classification, there is no need for separate cross-entropy loss. This is the reason why I removed it.

Sample-wise weights are more general than class-wise in a sense that for logloss we can model class-wise weights via sample-wise. This is the reason why I want to implement it in a more general way since it will cover both cases.

I agree that class_weights (or pos_weight) can be moved into the constructor. I'm ok with that.

About custom loss. The point was to create `TFFMBase` which can take any loss function. And `TFFMClassifier` and `TFFMRegressor` are just `TFFMBase` with the bundled loss (logloss and mse respectively). 
For now, right way to use some custom loss is to define your own class (similar to `TFFMClassifier`), define your loss and a way to make predictions (should it be normalized, sigmoid-transform and so on.)

peterewills commented:
Okay, that all makes sense. Was being thick about log-loss vs cross-entropy, I see it now. I'll add those keyword arguments to the constructor, and then add a demo of using weighted loss and custom loss into the examples notebook. Probably sometime next week. Thanks for all the input!

peterewills committed:
finished weighted loss examples in notebook

peterewills commented on tffm/base.py:
```diff
@@ -167,7 +167,10 @@ class TFFMBaseModel(six.with_metaclass(ABCMeta, BaseEstimator)):
+    def init_basemodel(self, n_epochs=100, batch_size=-1,
+                       log_dir=None,session_config=None,
+                       verbose=0, seed=None,sample_weight=None,
+                       pos_class_weight=None,**core_arguments):
```
made weights a parameter of the model so that they can be passed around from constructor to fit method

peterewills commented on tffm/base.py:
```diff
@@ -217,7 +222,7 @@ def _fit(self, X_, y_, w_, n_epochs=None, show_progress=False):
             epoch_loss = []
             # iterate over batches
-            for bX, bY, bW in batcher(X_[perm], y_=y_[perm], w_=[perm], batch_size=self.batch_size):
+            for bX, bY, bW in batcher(X_[perm], y_=y_[perm], w_=w_[perm], batch_size=self.batch_size):
```
this was a typo I think

peterewills commented on tffm/models.py:
```diff
@@ -27,19 +31,18 @@ def _preprocess_sample_weights(self, sample_weight, pos_class_weight, used_y):
-            used_w[used_y > 0] = 1.0 / pos_rate / 2
-            used_w[used_y < 0] = 1.0 / neg_rate / 2
+            used_w[used_y > 0] = neg_rate / pos_rate
+            used_w[used_y < 0] = 1.0
```
These SHOULD be equivalent, but they weren't. Since they cause the loss function to differ by a scaling factor, perhaps the effective learning rate changes or something like that. Anyways, the one I settled on ended up giving more reasonable performance.

peterewills reviewed (commented):
Made some changes to allow weights to be passed in constructor. Also added some examples to the example notebook, but more could be included if need be.

geffy commented:
Thank you for your effort!
Finally merged :)
